File: backend/schemas/workflow.py
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, validator
from datetime import datetime
from schemas.platform import PlatformType
from schema_node.email_val import EmailData
from schema_node.tele_val import TelegramData
from schema_node.trigger_val import TriggerData

class Node(BaseModel):
    id: str
    platform: PlatformType
    name: str
    data: Dict = {}
    credential_id: Optional[str] = None

    @validator("data")
    def validate_data(cls, v, values):
        platform = values.get("platform")

        if platform == PlatformType.TELEGRAM:
            return TelegramData(**v).dict() if v else {}

        elif platform == PlatformType.EMAIL:
            return EmailData(**v).dict() if v else {}

        elif platform == PlatformType.SLACK:
            return v  # Pass through for now

        elif platform == PlatformType.TRIGGER:
            return TriggerData(**v).dict() if v else {}

        return v  # Allow any data for unknown platforms


class Connection(BaseModel):
    source: str
    target: str

class WorkflowBase(BaseModel):
    title: str
    enabled: Optional[bool] = True
    nodes: Optional[List[Node]] = None
    connections: Optional[List[Connection]] = None

class WorkflowCreate(WorkflowBase):
    pass

class WorkflowUpdate(BaseModel):
    title: Optional[str] = None
    enabled: Optional[bool] = None
    nodes: Optional[List[Node]] = None
    connections: Optional[List[Connection]] = None

class WorkflowResponse(WorkflowBase):
    id: int
    user_id: int
    webhook_path: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    last_executed_at: Optional[datetime] = None

    class Config:
        from_attributes = True

File: frontend/src/types/workflow.ts
// TypeScript types matching backend schemas

export type PlatformType = 'telegram' | 'email' | 'slack' | 'trigger';

export interface Node {
  id: string;
  platform: PlatformType;
  name: string;
  data: Record<string, any>;
  credential_id?: string;
  // ReactFlow specific
  position?: { x: number; y: number };
  type?: string;
}

export interface Connection {
  source: string;
  target: string;
}

export interface Workflow {
  id: number;
  title: string;
  enabled: boolean;
  nodes: Node[] | null;
  connections: Connection[] | null;
  user_id: number;
  webhook_path?: string;
  created_at?: string;
  updated_at?: string;
  last_executed_at?: string;
}

export interface WorkflowCreate {
  title: string;
  enabled?: boolean;
  nodes?: Node[];
  connections?: Connection[];
}

export interface WorkflowUpdate {
  title?: string;
  enabled?: boolean;
  nodes?: Node[];
  connections?: Connection[];
}

export interface Credential {
  id: number;
  title: string;
  platform: PlatformType;
  data: Record<string, any>;
  user_id: number;
  created_at?: string;
}

export interface CredentialCreate {
  title: string;
  platform: PlatformType;
  data: Record<string, any>;
}

// Execution types
export interface ExecutionResult {
  workflow_id: number;
  webhook_path?: string;
  status: 'success' | 'failed';
  test_mode: boolean;
  execution_time_ms: number;
  result?: {
    status: string;
    executed_nodes: Array<{
      id: string;
      name: string;
      status: 'success' | 'skipped' | 'error';
      error?: string;
    }>;
    context: Record<string, any>;
  };
  error?: string;
}

// User types
export interface User {
  id: number;
  email: string;
  name?: string;
}

export interface AuthResponse {
  access_token: string;
  token_type: string;
  user: User;
}
